app/src/test/java/br/com/dcarv/arcmovies/data/tmdb/MockCountriesProvider.kt

```diff
@@ -0,0 +1,28 @@
+package br.com.dcarv.arcmovies.data.tmdb
+
+import br.com.dcarv.arcmovies.domain.abstraction.ICountriesListProvider
+import br.com.dcarv.arcmovies.domain.model.Country
+import io.reactivex.Observable
+
+/**
+ * Created by dfcarvalho on 20/01/18.
+ */
+class MockCountriesProvider : ICountriesListProvider {
```

Comment:
Yes, I have used Mockito before and it would definitely fit well here. I guess I wanted to show that this architecture lends itself well to unit tests even without the help of testing libraries.

The biggest problem were the activities and fragments, which were instantiating the presenters, providers, etc.. instead of getting them from DI or factories. That is one of the reasons I didn't write any test cases for the UI elements (the other reason was I'm not very familiar with Espresso but I'm trying to learn it). With Mockito, I could have used methods in the activities that returned these dependencies and then mocked in the tests.

Something like:

In MainActivity:
```Kotlin
fun getGenreProvider(): IMainPresenter {
    return TMDBGenreProvider(TMDB_API_KEY, service, mapper = TMDBGenreMapper)
}
```

In MainActivityTest:
```Kotlin
val mockedMainActivity = Mockito.mock(MainActivity.class)
val mockedGenreProvider = Mockito.mock(IGenreProvider.class)
// mock providers methods ...
mockedMainActivity.when(mockedMainActivity.getGenreProvider).thenReturn(mockedGenreProvider)
```

So yes, on a real app I would definitely use Mockito for unit testing, but I wanted to show (and see myself) what this architecture can do by itself.